File: next/kmp/core/src/commonMain/kotlin/org/dweb_browser/core/ipc/Ipc.kt
package org.dweb_browser.core.ipc

import io.ktor.http.Url
import kotlinx.atomicfu.atomic
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.mapNotNull
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.isActive
import kotlinx.coroutines.joinAll
import kotlinx.coroutines.launch
import kotlinx.coroutines.plus
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.dweb_browser.core.help.types.MicroModuleManifest
import org.dweb_browser.core.ipc.helper.EndpointIpcMessage
import org.dweb_browser.core.ipc.helper.IpcClientRequest
import org.dweb_browser.core.ipc.helper.IpcError
import org.dweb_browser.core.ipc.helper.IpcEvent
import org.dweb_browser.core.ipc.helper.IpcFork
import org.dweb_browser.core.ipc.helper.IpcLifecycle
import org.dweb_browser.core.ipc.helper.IpcLifecycleClosed
import org.dweb_browser.core.ipc.helper.IpcLifecycleClosing
import org.dweb_browser.core.ipc.helper.IpcLifecycleInit
import org.dweb_browser.core.ipc.helper.IpcLifecycleOpened
import org.dweb_browser.core.ipc.helper.IpcLifecycleOpening
import org.dweb_browser.core.ipc.helper.IpcMessage
import org.dweb_browser.core.ipc.helper.IpcRequest
import org.dweb_browser.core.ipc.helper.IpcResponse
import org.dweb_browser.core.ipc.helper.IpcServerRequest
import org.dweb_browser.core.ipc.helper.IpcStream
import org.dweb_browser.core.ipc.helper.toIpc
import org.dweb_browser.helper.Debugger
import org.dweb_browser.helper.DeferredSignal
import org.dweb_browser.helper.Producer
import org.dweb_browser.helper.SafeHashMap
import org.dweb_browser.helper.SafeLinkList
import org.dweb_browser.helper.SuspendOnce
import org.dweb_browser.helper.SuspendOnce1
import org.dweb_browser.helper.WARNING
import org.dweb_browser.helper.collectIn
import org.dweb_browser.helper.globalEmptyScope
import org.dweb_browser.helper.withScope
import org.dweb_browser.pure.http.IPureBody
import org.dweb_browser.pure.http.PureClientRequest
import org.dweb_browser.pure.http.PureHeaders
import org.dweb_browser.pure.http.PureMethod
import org.dweb_browser.pure.http.PureResponse


class Ipc internal constructor(
  val pid: Int,
  val endpoint: IpcEndpoint,
  val locale: MicroModuleManifest,
  val remote: MicroModuleManifest,
  val pool: IpcPool,
  val debugId: String = "${endpoint.debugId}/$pid",
) {
  val debugIpc by lazy { Debugger(this.toString()) }

  companion object {
    private val reqIdAcc = atomic(0)
  }

  private val job = SupervisorJob()
  val scope = endpoint.scope + job

  override fun toString() = "Ipc@$debugId"

  /**
   * 这部分得放最前面，因为有些地方需要立刻使用 onClosed
   */

  //#region close

  private val closeDeferred = CompletableDeferred<CancellationException?>()
  private val closeLock = Mutex(true)

  val isClosed get() = closeDeferred.isCompleted

  /**
   * 等待ipc关闭之后
   *
   * 对比 onBeforeClose ，该函数不在 ipc scope
   *
   * 这里之所以要放在closeLock，是因为 closeDeferred.await 和 invokeOnComplete 是同一个级别的。
   * 因此这里如果直接通过 closeDeferred.await 返回，那么 invokeOnComplete 可能没有执行完成，导致一些预期之外的效果
   */
  suspend fun awaitClosed() = closeLock.withLock { closeDeferred.await() }

  val onClosed = DeferredSignal(closeDeferred)

  // 开始触发关闭事件
  suspend fun close(cause: CancellationException? = null) {
    if (!isClosed) {
      closeOnce(cause)
    }
  }

  fun tryClose(cause: CancellationException? = null) {
    if (scope.isActive) {
      globalEmptyScope.launch(start = CoroutineStart.UNDISPATCHED) {
        close(cause)
      }
    }
  }

  /**
   * 长任务，需要全部完成才能结束ipcEndpoint
   */
  val launchJobs = SafeLinkList<Job>()

  private val closeOnce = SuspendOnce1 { cause: CancellationException? ->
    debugIpc("closing", cause)
    val reason = cause?.message
    IpcLifecycle(IpcLifecycleClosing(reason)).also { closing ->
      lifecycleLocaleFlow.emit(closing)
      sendLifecycleToRemote(closing)
    }
    messageProducer.producer.close(cause)
    closeDeferred.complete(cause)
    closeLock.unlock()
    IpcLifecycle(IpcLifecycleClosed(reason)).also { closed ->
      lifecycleLocaleFlow.emit(closed)
      runCatching { sendLifecycleToRemote(closed) }.getOrNull()
    }
    debugIpc.timeout(1000, "close") {
      launchJobs.joinAll()
    }
    scope.cancel(cause)
    debugIpc("closed", cause)
  }
  //#endregion

  /**
   * 消息生产者，所有的消息在这里分发出去
   */
  private val messageProducer = endpoint.getIpcMessageProducer(this)
  fun onMessage(name: String) = messageProducer.producer.consumer(name)

  //#region 生命周期相关的
  private val lifecycleLocaleFlow = MutableStateFlow(
    IpcLifecycle(IpcLifecycleInit(pid, locale.toCommonAppManifest(), remote.toCommonAppManifest()))
  )

  private val lifecycleRemoteFlow = onMessage("ipc-lifecycle-remote#$pid").mapNotNull { event ->
    event.consumeAs<IpcLifecycle>()
  }

  val lifecycle get() = lifecycleLocaleFlow.value
  val onLifecycle = lifecycleLocaleFlow.stateIn(scope, SharingStarted.Eagerly, lifecycle)


  // 标记ipc通道是否激活
  val isActivity get() = endpoint.isActivity

  suspend fun awaitOpen(reason: String? = null) = when (val state = lifecycle.state) {
    is IpcLifecycleOpened -> lifecycle
    else -> lifecycleLocaleFlow.mapNotNull {
      debugIpc("awaitOpen", "state=$it reason=$reason")
      val ipcLifecycle = when (it.state) {
        is IpcLifecycleOpened -> it
        is IpcLifecycleClosing, is IpcLifecycleClosed -> {
          throw IllegalStateException("ipc already closed")
        }

        else -> null
      }
      ipcLifecycle
    }.first().also {
      debugIpc("lifecycle-opened", reason)
    }
  }

  /**
   * 启动，会至少等到endpoint握手完成
   */
  suspend fun start(await: Boolean = true, reason: String = "") {
    if (await) {
      withScope(scope) {
        debugIpc.verbose("start-begin", reason)
        endpoint.start(true)
        startOnce()
        awaitOpen("from-start $reason")
      }
    } else {
      scope.launch {
        debugIpc.verbose("start-begin", reason)
        endpoint.start(true)
        startOnce()
      }
    }
  }

  private val startOnce = SuspendOnce {
    debugIpc("start", lifecycle)
    // 当前状态必须是从init开始
    when (val state = lifecycle.state) {
      // 告知对方我启动了
      is IpcLifecycleInit -> IpcLifecycle(IpcLifecycleOpening).also {
        sendLifecycleToRemote(it)
        debugIpc.verbose("emit-locale-lifecycle", it)
        lifecycleLocaleFlow.emit(it)
      }

      else -> throw IllegalStateException("fail to start: ipc=$this state=$state")
    }
    // 监听远端生命周期指令，进行协议协商
    lifecycleRemoteFlow.collectIn(scope) { lifecycleRemote ->
      debugIpc.verbose("lifecycle-in") { "remote=$lifecycleRemote locale=$lifecycle" }
      val doIpcOpened = suspend {
        IpcLifecycle(IpcLifecycleOpened).also {
          debugIpc.verbose("emit-locale-lifecycle", it)
          sendLifecycleToRemote(it)
          lifecycleLocaleFlow.emit(it)
        }
      }
      when (lifecycleRemote.state) {
        is IpcLifecycleClosing, is IpcLifecycleClosed -> tryClose(CancellationException("lifecycle close"))
        // 收到 opened 了，自己也设置成 opened，代表正式握手成功
        is IpcLifecycleOpened -> {
          when (lifecycle.state) {
            is IpcLifecycleOpening -> doIpcOpened()
            else -> {}
          }
        }
        // 如果对方是 init，代表刚刚初始化，那么发送目前自己的状态
        is IpcLifecycleInit -> sendLifecycleToRemote(lifecycle)
        // 等收到对方 Opening ，说明对方也开启了，那么开始协商协议，直到一致后才进入 Opened
        is IpcLifecycleOpening -> doIpcOpened()
      }
    }
    // 监听并分发 所有的消息
    onMessage("fork#$debugId").collectIn(scope) { event ->
      event.consumeAs<IpcFork> { ipcFork ->
        val forkedIpc = Ipc(
          pid = ipcFork.pid,
          endpoint = endpoint,
          locale = locale,
          remote = remote,
          pool = pool,
        )
        pool.safeCreatedIpc(
          forkedIpc, autoStart = ipcFork.autoStart, startReason = ipcFork.startReason
        )
        endpoint.forkedIpcMap.getOrPut(forkedIpc.pid) { CompletableDeferred() }.complete(forkedIpc)
        forkProducer.send(forkedIpc)
      }
    }
  }

  /**
   * 向远端发送 生命周期 信号
   */
  private suspend fun sendLifecycleToRemote(state: IpcLifecycle) {
    debugIpc.verbose("lifecycle-out", state)
    endpoint.postIpcMessage(EndpointIpcMessage(pid, state))
  }

  val waitForkedIpc = endpoint::waitForkedIpc

  /**
   * 在现有的线路中分叉出一个ipc通道
   * 如果自定义了 locale/remote，那么说明自己是帮别人代理
   */
  suspend fun fork(
    locale: MicroModuleManifest = this.locale,
    remote: MicroModuleManifest = this.remote,
    autoStart: Boolean = false,
    startReason: String? = null,
    pid: Int? = null,
  ): Ipc {
    awaitOpen("then-fork")
    // 这里确保 endpoint opened 了再 generatePid
    val ipcId = pid ?: endpoint.generatePid()
    val forkedIpc = pool.createIpc(
      pid = ipcId,
      endpoint = endpoint,
      locale = locale,
      remote = remote,
      autoStart = autoStart,
      startReason = startReason
    )
    endpoint.forkedIpcMap.getOrPut(forkedIpc.pid) { CompletableDeferred() }.complete(forkedIpc)
    // 自触发
    forkProducer.send(forkedIpc)
    // 通知对方
    postMessage(
      IpcFork(
        pid = forkedIpc.pid,
        autoStart = autoStart,
        startReason = startReason,
        /// 对调locale/remote
        locale = forkedIpc.remote.toCommonAppManifest(),
        remote = forkedIpc.locale.toCommonAppManifest(),
      )
    )
    return forkedIpc
  }

  private val forkProducer = Producer<Ipc>("fork#$debugId", scope)
  fun onFork(name: String) = forkProducer.consumer(name)

  //#endregion

  //#region 消息相关的


  private inline fun <T : Any> messagePipeMap(
    name: String,
    crossinline mapNotNull: suspend (value: IpcMessage) -> T?,
  ) = Producer<T>("$debugId/$name", scope).also { producer ->
    onClosed { cause ->
      producer.close(cause.exceptionOrNull())
    }
    onMessage(name).collectIn(scope) { event ->
      event.consumeMapNotNull {
        mapNotNull(it)
      }?.also {
        producer.send(it, event.order)
      }
    }
  }

  private val requestProducer by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
    messagePipeMap("request") { ipcMessage ->
      when (ipcMessage) {
        is IpcRequest -> when (ipcMessage) {
          is IpcClientRequest -> ipcMessage.toServer(this)
          is IpcServerRequest -> ipcMessage
        }

        else -> null
      }
    }
  }

  fun onRequest(name: String) = requestProducer.consumer(name)

  private val responseProducer by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
    messagePipeMap("response") {
      if (it is IpcResponse) it else null
    }
  }

  fun onResponse(name: String) = responseProducer.consumer(name)

  private val streamProducer by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
    messagePipeMap("stream") {
      if (it is IpcStream) it as IpcStream else null
    }
  }

  fun onStream(name: String) = streamProducer.consumer(name)

  private val eventProducer by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
    messagePipeMap("event") {
      if (it is IpcEvent) it else null
    }
  }

  fun onEvent(name: String) = eventProducer.consumer(name)


  private val errorProducer by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
    messagePipeMap("error") {
      if (it is IpcError) it else null
    }
  }

  fun onError(name: String) = errorProducer.consumer(name)

  suspend inline fun request(url: String) = request(
    PureClientRequest(
      method = PureMethod.GET,
      href = url,
    )
  )

  suspend inline fun request(url: Url) = request(url.toString())
  suspend fun request(ipcRequest: IpcRequest): IpcResponse {
    val result = CompletableDeferred<IpcResponse>()
    _reqResMap[ipcRequest.reqId] = result
    this.postMessage(ipcRequest)
    return result.await()
  }

  private val _reqResMap by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
    SafeHashMap<Int, CompletableDeferred<IpcResponse>>().also { reqResMap ->
      onResponse("req-res").collectIn(scope) { event ->
        // 消耗掉
        val response = event.consume()
        val result = reqResMap.remove(response.reqId) ?: return@collectIn debugIpc(
          "reqResMap",
          "onResponse",
          "no found response [${event.data}]  by reqId: ${event.data.reqId}"
        )
        result.complete(response)
      }
    }
  }

  private suspend inline fun buildIpcRequest(url: String, init: IpcRequestInit): IpcRequest {
    val reqId = this.allocReqId()
    return IpcClientRequest.fromRequest(reqId, this, url, init)
  }

  // PureClientRequest -> ipcRequest -> IpcResponse -> PureResponse
  suspend fun request(pureRequest: PureClientRequest): PureResponse {
    return request(
      pureRequest.toIpc(allocReqId(), this)
    ).toPure()
  }

  suspend fun request(url: String, init: IpcRequestInit): IpcResponse {
    val ipcRequest = this.buildIpcRequest(url, init)
    return request(ipcRequest)
  }

  private fun allocReqId() = reqIdAcc.incrementAndGet()


  /**发送各类消息到remote*/
  suspend fun postMessage(data: IpcMessage) {
    runCatching {
      awaitOpen("then-postMessage")
      withScope(scope) {
        endpoint.postIpcMessage(EndpointIpcMessage(pid, data))
      }
    }.getOrElse {
      WARNING("fail to postMessage: $data")
      WARNING(it.message)
    }
  }

  suspend inline fun postResponse(reqId: Int, response: PureResponse) {
    postMessage(IpcResponse.fromResponse(reqId, response, this))
  }
  //#endregion

  init {
    endpoint.onClosed {
      tryClose(CancellationException("endpoint closed", it))
    }
  }
}

data class IpcRequestInit(
  var method: PureMethod = PureMethod.GET,
  var body: IPureBody = IPureBody.Empty,
  var headers: PureHeaders = PureHeaders(),
)

File: toolkit/dweb-core/packages/js-core/src/ipc/ipc.ts
import { PromiseOut } from "../../helper/PromiseOut.ts";
import { CacheGetter } from "../../helper/cacheGetter.ts";
import type { $MicroModuleManifest } from "../types.ts";
import type { IpcHeaders } from "./helper/IpcHeaders.ts";

import {
  IpcClientRequest,
  IpcServerRequest,
} from "./ipc-message/IpcRequest.ts";
import { IpcResponse } from "./ipc-message/IpcResponse.ts";

import { $once, once } from "../../helper/decorator/$once.ts";
import { StateSignal } from "../../helper/StateSignal.ts";
import { CUSTOM_INSPECT, logger } from "../../helper/logger.ts";
import { mapHelper } from "../../helper/fun/mapHelper.ts";
import { promiseAsSignalListener } from "../../helper/promiseSignal.ts";
import { Producer } from "../../helper/Producer.js";
import { IpcPool } from "./IpcPool.ts";
import { endpointIpcMessage } from "./endpoint/EndpointIpcMessage.ts";
import { IpcEndpoint } from "./endpoint/IpcEndpoint.ts";
import { ipcFork } from "./ipc-message/IpcFork.ts";
import {
  ipcLifecycle,
  ipcLifecycleClosed,
  ipcLifecycleClosing,
  ipcLifecycleInit,
  ipcLifecycleOpened,
  ipcLifecycleOpening,
  type $IpcLifecycle,
} from "./ipc-message/IpcLifeCycle.ts";
import type { $IpcMessage } from "./ipc-message/IpcMessage.ts";
import { IPC_LIFECYCLE_STATE } from "./ipc-message/internal/IpcLifecycle.ts";
import { IPC_MESSAGE_TYPE } from "./ipc-message/internal/IpcMessage.ts";

export class Ipc {
  constructor(
    readonly pid: number,
    readonly endpoint: IpcEndpoint,
    readonly locale: $MicroModuleManifest,
    readonly remote: $MicroModuleManifest,
    readonly pool: IpcPool,
    readonly debugId = `${endpoint.debugId}/${pid}`
  ) {}
  toString() {
    return `Ipc#${this.debugId}`;
  }
  [CUSTOM_INSPECT]() {
    return this.toString();
  }
  readonly console = logger(this);

  // reqId计数
  #reqIdAcc = 0;
  // 消息生产者，所有的消息在这里分发出去
  #messageProducer = this.endpoint.getIpcMessageProducerByIpc(this);

  onMessage(name: string) {
    return this.#messageProducer.producer.consumer(name);
  }

  //#region 生命周期相关
  #lifecycleLocaleFlow = new StateSignal<$IpcLifecycle>(
    ipcLifecycle(ipcLifecycleInit(this.pid, this.locale, this.remote)),
    ipcLifecycle.equals
  );
  readonly lifecycleLocaleFlow = this.#lifecycleLocaleFlow.asReadyonly();
  get lifecycle() {
    return this.lifecycleLocaleFlow.state;
  }
  onLifecycle = this.lifecycleLocaleFlow.listen;

  #lifecycleRemoteFlow = this.onMessage(
    `ipc-lifecycle-remote#${this.pid}`
  ).mapNotNull((message) => {
    if (message.type === IPC_MESSAGE_TYPE.LIFECYCLE) {
      return message;
    }
  });
  readonly lifecycleRemoteFlow = this.#lifecycleRemoteFlow;
  /**
   * 向远端发送 生命周期 信号
   */
  #sendLifecycleToRemote(state: $IpcLifecycle) {
    this.console.debug("lifecycle-out", state);
    this.endpoint.postIpcMessage(endpointIpcMessage(this.pid, state));
  }

  // 标记ipc通道是否激活
  get isActivity() {
    return this.endpoint.isActivity;
  }

  /**等待启动 */
  async awaitOpen(reason?: string) {
    if (this.lifecycle.state.name === IPC_LIFECYCLE_STATE.OPENED) {
      return this.lifecycle;
    }
    const op = new PromiseOut<$IpcLifecycle>();
    const off = this.onLifecycle((lifecycle) => {
      switch (lifecycle.state.name) {
        case IPC_LIFECYCLE_STATE.OPENED: {
          op.resolve(lifecycle);
          break;
        }
        case (IPC_LIFECYCLE_STATE.CLOSED, IPC_LIFECYCLE_STATE.CLOSING): {
          op.reject("endpoint already closed");
          break;
        }
      }
    });
    const lifecycle = await op.promise;
    this.console.debug("awaitOpen", lifecycle, reason);
    off();
    return lifecycle;
  }

  /**
   * 启动，会至少等到endpoint握手完成
   */
  async start(isAwait = true, reason?: string) {
    this.console.debug("start", reason);
    if (isAwait) {
      this.endpoint.start(true);
      this.startOnce();
      await this.awaitOpen(`from-start ${reason}`);
    } else {
      this.endpoint.start(true);
      this.startOnce();
    }
  }

  startOnce = $once(() => {
    this.console.debug("startOnce", this.lifecycle);
    // 当前状态必须是从init开始
    if (this.lifecycle.state.name === IPC_LIFECYCLE_STATE.INIT) {
      // 告知对方我启动了
      const opening = ipcLifecycle(ipcLifecycleOpening());
      this.#sendLifecycleToRemote(opening);
      this.#lifecycleLocaleFlow.emit(opening);
    } else {
      throw new Error(`fail to start: ipc=${this} state=${this.lifecycle}`);
    }
    // 监听远端生命周期指令，进行协议协商
    this.#lifecycleRemoteFlow((lifecycleRemote) => {
      this.console.debug(
        "lifecycle-in",
        `remote=${lifecycleRemote},local=${this.lifecycle}`
      );
      // 告知启动完成
      const doIpcOpened = () => {
        const opend = ipcLifecycle(ipcLifecycleOpened());
        this.#sendLifecycleToRemote(opend);
        this.#lifecycleLocaleFlow.emit(opend);
      };
      // 处理远端生命周期
      switch (lifecycleRemote.state.name) {
        case (IPC_LIFECYCLE_STATE.CLOSING, IPC_LIFECYCLE_STATE.CLOSED): {
          this.close(lifecycleRemote.state.reason);
          break;
        }
        // 收到 opened 了，自己也设置成 opened，代表正式握手成功
        case IPC_LIFECYCLE_STATE.OPENED: {
          if (this.lifecycle.state.name === IPC_LIFECYCLE_STATE.OPENING) {
            doIpcOpened();
          }
          break;
        }
        // 如果对方是 init，代表刚刚初始化，那么发送目前自己的状态
        case IPC_LIFECYCLE_STATE.INIT: {
          this.#sendLifecycleToRemote(this.lifecycle);
          break;
        }
        // 等收到对方 Opening ，说明对方也开启了，那么开始协商协议，直到一致后才进入 Opened
        case IPC_LIFECYCLE_STATE.OPENING: {
          doIpcOpened();
          break;
        }
      }
    });
    // 监听并分发 所有的消息
    this.onMessage(`fork#${this.debugId}`).collect((event) => {
      const ipcFork = event.consumeMapNotNull((data) => {
        if (data.type === IPC_MESSAGE_TYPE.FORK) {
          return data;
        }
      });
      if (ipcFork === undefined) {
        return;
      }
      const forkedIpc = new Ipc(
        ipcFork.pid,
        this.endpoint,
        this.locale,
        this.remote,
        this.pool
      );
      this.pool.safeCreatedIpc(
        forkedIpc,
        ipcFork.autoStart,
        ipcFork.startReason
      );
      mapHelper
        .getOrPut(this.forkedIpcMap, forkedIpc.pid, () => new PromiseOut())
        .resolve(forkedIpc);
      this.#forkProducer.send(forkedIpc);
    });
  });
  //#region fork

  private forkedIpcMap = new Map<number, PromiseOut<Ipc>>();
  waitForkedIpc(pid: number) {
    return mapHelper.getOrPut(this.forkedIpcMap, pid, () => new PromiseOut())
      .promise;
  }

  /**
   * 在现有的线路中分叉出一个ipc通道
   * 如果自定义了 locale/remote，那么说明自己是帮别人代理
   */
  async fork(
    locale: $MicroModuleManifest = this.locale,
    remote: $MicroModuleManifest = this.remote,
    autoStart: boolean = false,
    startReason?: string
  ) {
    await this.awaitOpen("then-fork");
    const forkedIpc = this.pool.createIpc(
      this.endpoint,
      this.endpoint.generatePid(),
      locale,
      remote,
      autoStart,
      startReason
    );
    mapHelper
      .getOrPut(this.forkedIpcMap, forkedIpc.pid, () => new PromiseOut())
      .resolve(forkedIpc);
    // 自触发
    this.#forkProducer.send(forkedIpc);
    // 通知对方
    this.postMessage(
      ipcFork(
        forkedIpc.pid,
        autoStart,
        /// 对调locale/remote
        forkedIpc.remote,
        forkedIpc.locale,
        startReason
      )
    );
    return forkedIpc;
  }

  #forkProducer = new Producer<Ipc>(`fork#${this.debugId}`);
  onFork(name: string) {
    return this.#forkProducer.consumer(name);
  }
  //#endregion

  //#region 消息相关的
  #messagePipeMap<R>(
    name: string,
    mapNotNull: (value: $IpcMessage) => R | undefined
  ) {
    const producer = new Producer<R>(
      this.#messageProducer.producer.name + "/" + name
    );
    this.onClosed((reason) => {
      return producer.close(reason);
    });
    const consumer = this.onMessage(name);
    consumer.collect((event) => {
      const result = event.consumeMapNotNull<R>(mapNotNull);
      if (result === undefined) {
        return;
      }
      producer.send(result);
    });
    producer.onClosed(() => {
      consumer.cancel();
    });
    return producer;
  }
  #requestProducer = new CacheGetter(() =>
    this.#messagePipeMap("request", (ipcMessage) => {
      if (ipcMessage instanceof IpcClientRequest) {
        return ipcMessage.toServer(this);
      } else if (ipcMessage instanceof IpcServerRequest) {
        return ipcMessage;
      }
    })
  );
  onRequest(name: string) {
    return this.#requestProducer.value.consumer(name);
  }
  #responseProducer = new CacheGetter(() =>
    this.#messagePipeMap("response", (ipcMessage) => {
      if (ipcMessage instanceof IpcResponse) {
        return ipcMessage;
      }
    })
  );
  onResponse(name: string) {
    return this.#responseProducer.value.consumer(name);
  }
  #streamProducer = new CacheGetter(() =>
    this.#messagePipeMap("stream", (ipcMessage) => {
      if ("stream_id" in ipcMessage) {
        return ipcMessage;
      }
    })
  );
  onStream(name: string) {
    return this.#streamProducer.value.consumer(name);
  }
  #eventProducer = new CacheGetter(() =>
    this.#messagePipeMap("event", (ipcMessage) => {
      if (ipcMessage.type === IPC_MESSAGE_TYPE.EVENT) {
        return ipcMessage;
      }
    })
  );
  onEvent(name: string) {
    return this.#eventProducer.value.consumer(name);
  }
  #errorProducer = new CacheGetter(() =>
    this.#messagePipeMap("error", (ipcMessage) => {
      if (ipcMessage.type === IPC_MESSAGE_TYPE.ERROR) {
        return ipcMessage;
      }
    })
  );
  onError(name: string) {
    return this.#errorProducer.value.consumer(name);
  }

  #reqResMap = new CacheGetter(() => {
    const reqResMap = new Map<number, PromiseOut<IpcResponse>>();
    this.onResponse("req-res").collect((event) => {
      const response = event.consume();
      const result = mapHelper.getAndRemove(reqResMap, response.reqId);
      if (result === undefined) {
        throw new Error(`no found response by reqId: ${event.data.reqId}`);
      }
      result.resolve(response);
    });
    return reqResMap;
  });

  /** 发起请求并等待响应 */
  request(url: IpcClientRequest): Promise<IpcResponse>;
  request(url: string, init?: $IpcRequestInit): Promise<IpcResponse>;
  request(input: string | IpcClientRequest, init?: $IpcRequestInit) {
    const ipcRequest =
      input instanceof IpcClientRequest
        ? input
        : this.#buildIpcRequest(input, init);
    const result = this.#registerReqId(ipcRequest.reqId);
    this.postMessage(ipcRequest);
    return result.promise;
  }
  /** 自定义注册 请求与响应 的id */
  #registerReqId(reqId = this.#allocReqId()) {
    return mapHelper.getOrPut(
      this.#reqResMap.value,
      reqId,
      () => new PromiseOut()
    );
  }
  #buildIpcRequest(url: string, init?: $IpcRequestInit) {
    const reqId = this.#allocReqId();
    const ipcRequest = IpcClientRequest.fromRequest(reqId, this, url, init);
    return ipcRequest;
  }
  #allocReqId() {
    return this.#reqIdAcc++;
  }

  //#endregion

  async postMessage(message: $IpcMessage) {
    try {
      await this.awaitOpen("then-postMessage");
    } catch (e) {
      this.console.debug(`ipc(${this}) fail to poseMessage: ${e}`);
      return;
    }
    this.endpoint.postIpcMessage(endpointIpcMessage(this.pid, message));
  }

  //#endregion

  //#region close

  @once()
  private get _closePo() {
    return new PromiseOut<string | undefined>();
  }
  awaitClosed() {
    return this._closePo.promise;
  }
  @once()
  get onClosed() {
    return promiseAsSignalListener(this._closePo.promise);
  }
  get isClosed() {
    return this.lifecycle.state.name == IPC_LIFECYCLE_STATE.CLOSED;
  }

  #closeOnce = $once(async (cause?: string) => {
    this.console.debug("closing", cause);
    {
      const closing = ipcLifecycle(ipcLifecycleClosing(cause));
      this.#lifecycleLocaleFlow.emit(closing);
      this.#sendLifecycleToRemote(closing);
    }
    await this.#messageProducer.producer.close(cause);
    this._closePo.resolve(cause);
    {
      const closed = ipcLifecycle(ipcLifecycleClosed(cause));
      this.#lifecycleLocaleFlow.emitAndClear(closed);
      this.#sendLifecycleToRemote(closed);
    }
  });

  async close(cause?: string) {
    this.#closeOnce(cause);
    this.#destroy();
  }

  /**销毁ipc */
  private _isDestroy = false;
  async #destroy() {}

  /**----- close end*/
}
export type $IpcRequestInit = {
  method?: string;
  body?: /* json+text */
  | null
    | string
    /* base64 */
    | ArrayBuffer
    | ArrayBufferView
    /* stream+base64 */
    | Blob
    | ReadableStream<Uint8Array>;
  headers?: IpcHeaders | HeadersInit;
};
